File: lecture_5/p5.java
public class p5 {

    public static int kSubsets(int n,int k)
    {
        if(n==0 || k==0 || k>n)
            return 0;
        else if(n==1 || k==1)
            return 1;
        int count=0;
        count=(n-1)*kSubsets(n-1,k);
        count+=kSubsets(n-1,k-1);
        return count;
    }

    public static void main(String args[])
    {
        int n=3;
        int k=2;
        System.out.println(kSubsets(n,k));
    }
}

File: lecture_39/p5.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set <string> wlist;
        for(string s:wordList)
            wlist.insert(s);
        queue<string> que;
        int l=1;
        que.push(beginWord);
        wlist.erase(beginWord);
        while(que.size()>0)
        {
            int size=que.size();
            while(size-->0)
            {
                string rs=que.front();
                que.pop();
                for(int i=0;i<rs.length();i++)
                {
                    char oc=rs[i];
                    for(int j=0;j<26;j++)
                    {
                        char c='a'+j;
                        rs[i]=c;
                        //if(wlist.find(rs)!=wlist.end())     // slow
                        if(wlist.count(rs)!=0)
                        {
                            if(rs==endWord)
                                return l+1;
                            que.push(rs);
                            wlist.erase(rs);
                        }
                    }
                    rs[i]=oc;
                }
            }
            l++;
        }
        return 0;
    }
};